### PR TITLE
Download cRAP

### PR DESCRIPTION
**Pull request by csdaw**
Adds the following functions (with appropriate tests and examples):

- `download_crap()` wrapper function to download different types of contaminants databases. For now the only type available is `type = "ccp"` which will call the `download_ccp_crap()` function. Easily extendable in the future.
- `download_ccp_crap()`, an internal function that will download an up-to-date version of the CCP cRAP database from the latest UniProt release
- `check_uniprot_release()` just returns the latest UniProt release as a character e.g. `"2021_01"`. I use this function when naming files I've downloaded
- `sub_crap()` just adds cRAP numbers to a character vector e.g. turns `sp|XXX|YYY` into `sp|cRAP001|XXX|YYY`

Adds the following vignette:

- `crap.Rmd` = brief discussion of contaminant databases, then shows how to use `download_crap()` and also how to use `Biostrings` to add your own sequences of interest to a cRAP FASTA file.

### CODE REVIEW

**Review by TomSmithCGAT — COMMENTED** [review]
A very well explained vignette. I think we might have some differences of opinion in exactly how to split out the functionality but if you decide to ignore my thoughts on that side, I've got no issues with the functions as they stand :+1:

File: R/download_crap.R [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+#' Default is `TRUE`
+#'
+#' @return Returns a file saved to disk at the specified filepath
+#' @examples
+#' \dontrun{
+#' download_crap("path/to/file/crap_2021_01.fasta", type = "ccp")
+#' }
+#'
+#' @export
+download_crap <- function(file, type = "ccp", add_crap = TRUE,
```
[the remark this answers is not in the document]
Comment (csdaw) [reply R/download_crap.R:58]: I think I agree with you. The only difficulty would be when you want sequences in your crap fasta that aren't in UniProt.

I can envision 3 functions: 

`get_ccp_crap()` which outputs the accessions from CCP cRAP (minus the commercial sequences which aren't in any database).

`make_crap_fasta()` which takes a vector of UniProt accessions, downloads the sequences,  adds `cRAP00x` to the headers, and outputs a fasta file. 

`append_crap_fasta()` which takes an existing fasta (e.g. contains commercial sequences), adds `cRAP00x` to the headers, and sticks it on the end of an existing crap fasta.

Then you would have `download_ccp_crap()` wrapper function to envelop these 3 functions.

File: vignettes/crap.Rmd [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+## Introduction
+
+Peptide-spectrum matching requires 2 main inputs: a collection of MS2 spectra 
+and a FASTA database of protein/peptide sequences. Search algorithms are then
+used to match the supplied MS2 spectra to theoretical spectra generated from
+an in-silico digest of the supplied FASTA database (typically we use the 
+SwissProt protein database for the organism of interest). The selection of
+the FASTA database should be done carefully as it can have a 
+[significant effect on protein quantification](https://pwilmart.github.io/blog/2020/09/19/shotgun-quantification-part2)
```
[the remark this answers is not in the document]
Comment (csdaw) [reply vignettes/crap.Rmd:34]: I found a short section (Impact of the size of the database) in this paper where they mention increasing FASTA size decreases number of PSMs https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-13-S16-S2

[the remark this answers is not in the document]
Comment (TomSmithCGAT) [reply vignettes/crap.Rmd:34]: Great :+1: Makes perfect sense but nice to have a proper citation for that

**Comment by csdaw** [comment]
This PR now adds the following functions:
- `make_crap_fasta()` which takes a character vector of UniProt accessions and a file path. It queries UniProt for the sequences and saves them into a fasta at the specified file path (with appropriate `cRAP00X` numbers if specified).
- `append_crap_fasta()` which takes 2 file paths: one of the fasta to add, and another of the existing cRAP fasta to append to. It adds the sequences in the first fasta to the end of the second (with appropriate `cRAP00X` numbers if specified).
- `get_ccp_crap()` which does not take any inputs and outputs a character vector of CCP cRAP UniProt sequences.
- `download_ccp_crap()` wrapper function, input is a file path and the function downloads some sequences and saves it to the specified file path. Wraps `get_ccp_crap()`, `make_crap_fasta()`, and `append_crap_fasta()` in that order.
- - `check_uniprot_release()` just returns the latest UniProt release as a character e.g. `"2021_01"`. I use this function when naming files I've downloaded
- `sub_crap()` just adds cRAP numbers to a character vector e.g. turns `sp|XXX|YYY` into `sp|cRAP001|XXX|YYY`

Adds the following vignette:

- `crap.Rmd` = brief discussion of contaminant databases, then shows how to use `download_ccp_crap()`, `make_crap_fasta()`, and `append_crap_fasta()`.

**Comment by TomSmithCGAT** [comment]
Brilliant. If I may continue my suggestions, `make_crap_fasta)` can be used to generate a fasta for any set of accessions, right? Would this be better called `make_fasta()`, with argument `is_crap` replacing `add_crap` to specify that the entry names should be renamed to reflect their crappy nature. Ditto `append_crap_fasta()`.

This would make it easy to add future functions to identify accessions for e.g all human swissProt, from which a fasta file could be generated using `make_fasta()`, which may also later need an argument to reformat fasta entries to make them the expect format for e.g PD fasta parsing.

**Comment by csdaw** [comment]
Yes `is_crap` makes sense to me!

I don't quite understand your second point though.

**Comment by TomSmithCGAT** [comment]
Right now, your functions are written with crap proteins in mind. But with `is_crap=F`, they are applicable to any set of uniprot accessions. By making that explicit in the function names and argument, it's clearer that they can also be used if one wants to make a bespoke reference fasta e.g all human SwissProt proteins + transgenes etc

**Comment by csdaw** [comment]
Ah I see what you mean.

I'll need to make it explicit in the `make_fasta()` docs that it can only be used for so many accessions at once time. To quote UniProt:

> Very large mapping requests (>50,000 identifiers) are likely to fail. Please do verify that your list does not contain any duplicates, and try to split it into smaller chunks (<20,000) in case of problems.

Probably it would be good to write a function called `download_proteome(id = "UP000000xxx", isoforms = FALSE, swissprot_only = TRUE)` which would use a different httr query to download UniProt's already put together FASTA files which exist for many (but not all) species. I have a script that could be adapted for this.

**Comment by TomSmithCGAT** [comment]
Ah, yes, good points. Should definitely download reference fastas where they are available, as you suggest.

For now, I think it's fine just to rename functions & arguments so the above can be merged.

**Comment by csdaw** [comment]
Alright I think this is it.

This PR now adds the following functions:
- `make_fasta()` which takes a character vector of UniProt accessions and a file path. It queries UniProt for the sequences and saves them into a fasta at the specified file path (with appropriate `cRAP00X` numbers if `is_crap = TRUE`).
- `append_fasta()` which takes 2 file paths: file1 = the fasta to append and file2 = the fasta to append to. It adds the sequences in the file1 (with appropriate `cRAP00X` numbers if `is_crap = TRUE`) to the end of file2 .
- `get_ccp_crap()` which does not take any inputs and outputs a character vector of CCP cRAP UniProt sequences.
- `download_ccp_crap()` wrapper function, input is a file path and the function downloads some sequences and saves it to the specified file path. Wraps `get_ccp_crap()`, `make_fasta()`, and `append_fasta()` in that order.
- - `check_uniprot_release()` just returns the latest UniProt release as a character e.g. `"2021_01"`. I use this function when naming files I've downloaded
- `sub_crap()` just adds cRAP numbers to a character vector e.g. turns `sp|XXX|YYY` into `sp|cRAP001|XXX|YYY`

Adds the following vignette:

- `crap.Rmd` = brief discussion of contaminant databases, then shows how to use `download_ccp_crap()`, `make_fasta()`, and `append_fasta()`.

**Comment by TomSmithCGAT** [comment]
Great. Nothing more from me :sweat_smile:

**Pull request merged**